### DISCUSSION TITLE
How to create a new/custom keyboard layout?

### DISCUSSION THREAD

**Discussion by Savasdotexe** [AnySoftKeyboard/AnySoftKeyboard#3660] [General]
Using the 'English, QWERTY Latin keyboard', I'd like to change the emoji button to an apostrophe or maybe other designs.

I can't find anything on the Wiki that explains how to do this, also someone asked (#1529) for a manual on this, but that was back in 2018 so I'm guessing more people have asked, but I can't simply find the solution. Various suggestions, but unclear on exactly what needs to be done.

**Comment by nicoursi** [comment]
You  can find useful Info here:
https://github.com/AnySoftKeyboard/AnySoftKeyboard/blob/main/addons/CONTRIBUTING.md

Another good trick is to look at some merged PRs commits so you can understand what files you need to change/create.


Some examples:
Generic rows (top rows and bottom rows):
#2148

Keyboard layouts:
#2235 #2149 #2150

**Reply by khlsvr** [reply]
Is there a simple way for me to just change one default long-press key in one layout or do I need to go through a process of installing Android Studio and what not? I mainly type english and finnish, so I need the `ä` character a lot. I also use dvorak layout. There are 2 different dvorak layouts available. One where long-pressing a gives a with \` on top of it and the other one giving @. I would want the long-press to produce `ä`. Well, I could get used to being able to get the `ä` by long-pressing whatever key on the keyboard if that's a problem. Currently I need to long-press `a` and slide sideways to pick `ä`.

**Reply by nicoursi** [reply]
Setting up a building environment is recommended so that you can build and test your PR, otherwise you wouldn't know if the language pack you coded works as expected or not.

Android studio needs to be installed but for building you'd use the gradle scripts. Building from withing Android Studio will fail.
See:. https://github.com/AnySoftKeyboard/AnySoftKeyboard/blob/main/CONTRIBUTING.md (This info is a bit outdated. Just install the latest Android Studio and latest NDK. follow the instructions and after closing the studio try to build the Finnish pack, without modifications, to allow Gradle to install all missing components→`./gradlew :addons:languages:finnish:apk:assembleDebug`) 

You'll also need to install and setup `git`, so that you can push from shell.

Alternatively, but not recommended, you could just fork, create a branch, and add new code just from the web interface of GitHub.

Regarding the layout, what I would suggest is to add two new layouts to the Finnish pack (Dvorak and qwerty), just by copying the English layouts you like the most and applying the modifications you suggested above.
